enforcer.go

```diff
@@ -419,78 +423,76 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
 				len(rvals),
 				rvals)
 		}
-		for i, pvals := range e.model["p"]["p"].Policy {
-			// log.LogPrint("Policy Rule: ", pvals)
-			if len(e.model["p"]["p"].Tokens) != len(pvals) {
-				return false, fmt.Errorf(
-					"invalid policy size: expected %d, got %d, pvals: %v",
-					len(e.model["p"]["p"].Tokens),
-					len(pvals),
-					pvals)
+
+		if e.parallel {
+			ctx, cancel := context.WithCancel(context.Background())
+
+			n := runtime.GOMAXPROCS(0)
+
+			var wg sync.WaitGroup
+			wg.Add(n)
+
+			type param struct {
+				i          int
+				parameters enforceParameters
+				expression *govaluate.EvaluableExpression
 			}
 
-			parameters.pVals = pvals
-
-			if hasEval {
-				ruleNames := util.GetEvalValue(expString)
-				var expWithRule = expString
-				for _, ruleName := range ruleNames {
-					if j, ok := parameters.pTokens[ruleName]; ok {
-						rule := util.EscapeAssertion(pvals[j])
-						expWithRule = util.ReplaceEval(expWithRule, rule)
-					} else {
-						return false, errors.New("please make sure rule exists in policy when using eval() in matcher")
+			inputs := make(chan param, n)
+			errs := make(chan error, len(e.model["p"]["p"].Policy))
+
+			for i := 0; i < n; i++ {
+				go func() {
+					defer wg.Done()
+
+					for input := range inputs {
```

Comment:
I don't think this is correct. You create N goroutine to wait data from other goroutine, but we only have a datasource. N goroutine to do the same thing.  What do you think?

Reply:
The following is my implement:
```golang
n := runtime.GOMAXPROCS(0)
jobC := make(chan struct{}, n)
for i, pvals := range e.model["p"]["p"].Policy {
	jobC <- struct{}{}

	go func(done chan struct{}) {
		defer func() {
			<-done
		}()
		// do eval policy
	}(jobC)
}
````